feat: [Breaking] web form refactor

scmmishra opened the pull request:
## Changes
- Removed the old scripts, replaced with new `web_form` class and `web_form_list` class.
- The old API used in custom scripts are still compatible
- [Breaking] Removed `page_break` field type
- Removed some tests, need to replace with cypress tests
- New list view filter, use wildcard search for data field.
- Page doesn't reload after applying filter in list view
- Fixed rendering issue on most fields inside webforms

scmmishra committed:
fix [web-forms]: fix add_new_row for child table

- add df.data to this in add_new_row when frm not available
- fixed get_data function to send df.data if frm not available
- modified toggle for empty state

scmmishra committed:
fix [web-form]: Used hidden css class instead of hide

hidden class is available in both website and desk, hide is not.

scmmishra committed:
fix: added null check for locals in frappe.get_meta

Since locals object is not populated for website, frappe.get_meta would throw an exception.
This would break grid row in web forms.

scmmishra committed:
chore: removed commented code

scmmishra committed:
fix: delete row in grid

scmmishra committed:
fix: Added octicons to webforms

scmmishra committed:
fix: Layout of grids in webform

scmmishra committed:
fix [upload]: Added vue to webform

scmmishra committed:
fix: Modal UI

scmmishra committed:
style: linting fixes

scmmishra committed:
chore: Added undefined check for frappe.boot.user in datepicker

scmmishra committed:
feat: Added webform.js and new WebForm class

scmmishra committed:
feat: Added new WebForm classes to build.json

scmmishra committed:
feat: modified webform template

scmmishra committed:
chore: css fixes for web_form template

scmmishra committed:
feat: Added web_form check in render.py

To-Do:
- Refactor get_context web forms

scmmishra committed:
chore: Minor change to web_form.js

scmmishra committed:
feat: Added save function to web_form

scmmishra committed:
style: better code formatting

scmmishra committed:
feat: WebForm save implemented

scmmishra committed:
feat: Forms now allow editing of existing documents

scmmishra committed:
feat: Configured list view for web forms

scmmishra committed:
feat: Added WebFormList class

scmmishra committed:
feat: Added WebFormListRow class

scmmishra committed:
feat: Web form allows update and save

scmmishra committed:
chore: added styles and other linting fixes

scmmishra committed:
feat: Added 'Show More' feature to webform list

scmmishra committed:
chore: styling and navigation changes

scmmishra committed:
chore: moved webform list class to a separate file

scmmishra committed:
chore: added field formatter

scmmishra committed:
chore: styling table

scmmishra committed:
feat: Added web-form filters

scmmishra committed:
chore: replaced ast.literal_eval with json.loads in prepare filter

scmmishra committed:
chore: styling fixes for table in web_form list

scmmishra committed:
chore: allow guest only if login is not required

scmmishra committed:
chore: modified link.js to work on web forms

scmmishra committed:
fix: link fields in webforms

scmmishra committed:
style: removed stray console.log

scmmishra committed:
Merge branch 'develop' of https://github.com/frappe/frappe into web-form-refactor

netchampfaris commented on frappe/public/js/frappe/form/controls/link.js:
```diff
@@ -52,7 +52,13 @@ frappe.ui.form.ControlLink = frappe.ui.form.ControlData.extend({
+		if (this.doctype) return this.doctype
+		else {
+			if(typeof frappe.get_route === "function") {
+				return frappe.get_route()[0] === 'List' ? frappe.get_route()[1] : null;
```
Simple existence check?

```suggestion
				return frappe.get_route && frappe.get_route()[0] === 'List' ? frappe.get_route()[1] : null;
```

netchampfaris commented on frappe/public/js/frappe/form/grid.js:
```diff
@@ -136,19 +136,24 @@ export default class Grid {
 			me.get_selected().forEach((docname) => {
 				tasks.push(() => {
 					if (!me.frm) {
-						me.deleted_docs.push(docname);
```
Why remove `deleted_docs`? Isn't it used somewhere?

scmmishra replied:
In my testing I didn't see it being used anywhere except for this place where it was broken anyway. 
In this particular case, the `if` block it is in, only works for webforms and not on the desk. So event in the worst case scenario, the desk stays intact.

netchampfaris commented on frappe/public/js/frappe/web_form/web_form_class.js:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	setup_delete_button() {
+		const delete_button = document.createElement("button");
```
Why avoid jQuery? It's not going anywhere anytime soon..

scmmishra replied:
No particular reason, had slightly better experience with native API and it's mostly inter-operable.

netchampfaris commented on frappe/public/js/frappe/web_form/web_form_list_class.js:
```diff
@@ -0,0 +1,311 @@
+	}
+};
+
+frappe.ui.WebFromListRow = class WebFromListRow {
```
```suggestion
frappe.ui.WebFormListRow = class WebFormListRow {
```

scmmishra committed:
chore: removed old webform scripts

scmmishra committed:
style: fix minor typo

scmmishra committed:
refactor: simpler null check in get_reference_doctype

scmmishra committed:
fix: add fields to Table df in webform

scmmishra committed:
refactor: Use this.df if frappe.meta returns null in grid.js

scmmishra committed:
fix: null check for field in formatter

scmmishra committed:
feat: Web form custom scripts now working

scmmishra committed:
fix: fix add new row bug

scmmishra committed:
fix: grids in webform

scmmishra committed:
fix: links in webform girds

scmmishra committed:
fix: grid row form in webforms

scmmishra committed:
chore: added comments to webforms

scmmishra committed:
chore: added print button

scmmishra committed:
chore: added attachments

scmmishra committed:
fix: checkbox not getting unchecked on refresh

scmmishra committed:
chore: added form description

scmmishra committed:
refactor: minor code improvements

scmmishra committed:
Merge branch 'develop' of https://github.com/frappe/frappe into web-form-refactor

scmmishra committed:
style: improved readability

scmmishra committed:
fix: allow incomplete forms

scmmishra committed:
fix: allow payments in web forms

scmmishra committed:
fix: redirect to bug in razorpay

scmmishra committed:
refactor: remove page break field for webforms

scmmishra committed:
tests: modified test for new webforms

scmmishra committed:
style: linting fixes

scmmishra committed:
style (linting): added semi-colon

scmmishra committed:
test: remove old test

to be replaced with cypress

scmmishra committed:
refactor: renamed web_form scripts

old                   >         new
web_form.js               webform_script.js
web_form_class.js         web_form.js
web_form_class_list.js    web_form_list.js

scmmishra committed:
fix: select field not working in web_form

scmmishra committed:
fix: read_only fields in web_forms

scmmishra committed:
refactor: added webform list to frappe.ui namespace

scmmishra committed:
Merge branch 'develop' of https://github.com/frappe/frappe into web-form-refactor

scmmishra committed:
fix: font-size for web-form grids

scmmishra committed:
style: linting fixes

netchampfaris committed:
fix: Remove duplicated css

netchampfaris committed:
fix: Parse for_payment

netchampfaris committed:
fix: Web form list view fields

netchampfaris committed:
fix: Move quill.less to controls.less

netchampfaris committed:
fix: Minor fixes

netchampfaris committed:
fix: Remove jquery plugins

netchampfaris committed:
fix: Refactor into import statements

netchampfaris committed:
fix: Fix locals check

netchampfaris committed:
fix: Attachments

netchampfaris committed:
style: Unnecessary semicolon

netchampfaris committed:
fix: Trigger change handler onchange

netchampfaris committed:
refactor: web_form.html and web_form_script

- Remove web_form_settings
- Reorganize template for list and form
- Run web_form script only in form view

netchampfaris committed:
style: Missing semicolon
